doc/source/raysgd/v2/migration-guide.rst

```diff
@@ -0,0 +1,268 @@
+           add_dist_sampler=True
+
+       training_operator.setup(config)
+
+       for idx in range(config["num_epochs"]):
+           train_loader = training_operator._get_train_loader()
+           # If using DistributedSampler, set the epoch here.
+           train_loader.set_epoch(idx)
+           training_operator.train_epoch(epoch_idx=idx, iter(train_loader))
+
+           validation_loader = training_operator._get_validation_loader()
+           training_operator.validate(iterator=iter(validation_loader))
+
+       if sgd.world_rank() == 0:
+           return training_operator._get_original_models()
+       else:
+           return None
+
+   trainer = Trainer(backend="torch", num_workers=4, use_gpu=True)
+   trainer.start()
+   results = trainer.run(train_func, config={"num_epochs": 10})
+   final_model = results[0]
+
+Tensorflow
+~~~~~~~~~~
+
+The API for ``TFTrainer`` uses creator functions instead of a ``TrainingOperator`` to define the training logic. To port over Ray SGD v1 Tensorflow code to v2 you can do the following:
+
+.. code-block:: python
+
+   def train_func(config):
+       train_dataset, val_dataset = data_creator(config)
+       strategy = MultiWorkerMirroredStrategy()
+       with strategy.scope():
+           model = model_creator(config)
+
+       for epoch_idx in range(config["num_epochs"]):
+           model.fit(train_dataset)
+
+       if sgd.world_rank() == 0:
+           return model
+       else:
+           return None
+
+   trainer = Trainer(backend="tensorflow", num_workers=4, config={"num_epochs": 3, ...})
+   trainer.start()
+   model = trainer.run(train_func)[0]
+
+
+.. _sgd-migration-trainer:
+
+Interacting with the ``Trainer``
+--------------------------------
+
+In Ray SGD v1, you can iteratively call ``trainer.train()`` or ``trainer.validate()`` for each epoch, and can then interact with the trainer to get certain state (model, checkpoints, results, etc.). In Ray SGD v2, this is replaced by a single training function that defines the full training & validation loop for all epochs.
+
+There are 3 ways to get state during or after the training execution:
+
+
+#. Return values from your training function
+#. Intermediate results via ``sgd.report()``
+#. Saving & loading checkpoints via ``sgd.save_checkpoint()`` and ``sgd.load_checkpoint()``
+
+Return Values
+~~~~~~~~~~~~~
+
+To get any state from training *after* training has completed, you can simply return it from your training function. The return values from each the workers will be added to a list and returned from the ``trainer.run()`` call.
+
+.. code-block:: python
+
+   def train_func():
+       model = Net()
+       trainer_loader = MyDataset()
+       for batch in train_loader:
+           model.train(batch)
+
+       return model
+
+   trainer = Trainer(backend="torch")
+   trainer.start()
+   results = trainer.run(train_func, num_workers=2)
+   assert len(results) == 2
+   trained_model = results[0]
+
+Intermediate Reporting
```

Comment:
For these sections do you think it'd be helpful to include "before" API code (how to do it with SGD v1)? That way it is easy for the user to know exactly which section to look at for any given existing functionality.

Reply:
Done